### ISSUE TITLE
Call to `previous_model` is triggering `transaction_changed_attributes` callbacks causing changes to be overwritten

### ISSUE THREAD

**Issue by jfloff** [magnusvk/counter_culture#384]
Hi there,

I'm reporting an issue here after observing potential complications with the `previous_model` call. It seems that this call has caused other problems before, and I wanted to bring it to your attention. Additionally, I'm reaching out to @dylanahsmith, the author of the [ArTransactionChanges](https://github.com/dylanahsmith/ar_transaction_changes/) gem, to gather his insights on this matter.

As you may know, the `counter_culture` gem utilizes a [combination of the `dup` call and direct assignment of previous changes](https://github.com/magnusvk/counter_culture/blob/74da0105539904a935f8251a03b9bb25ba99aac5/lib/counter_culture/counter.rb#L311) to track changes to the monitored attributes. These changes are later used to compute differences applied to the counter attributes in the parent model.

The issue here arises from duplicating an ActiveRecord model and then assigning new values to the tracked attributes. This triggers gems that intercept the [write_attribute](https://apidock.com/rails/ActiveRecord/AttributeMethods/write_attribute) call, such as ArTransactionChanges. Specifically, ArTransactionChanges detects a change but removes it from the set of changes recorded at the end. For example, if I set the quantity from 0 to 1, `transaction_changed_attributes` tracks that change. However, due to the subsequent `dup` & assign call, there is an action that negates the effects of the first one, resulting in the removal of that change from the set.

In conclusion, I'm questioning whether the strategy of `dup` and assign is the most suitable one. While it currently causes issues with the ArTransactionChanges gem, I believe there may be similar problems with other gems that intercept the `write_attribute` behavior.

Your attention to this matter is much appreciated.

Thank you.

**Comment by jfloff** [comment]
I've made an attempt to fix the problem. Maybe better solutions are out there with more structural changes, but this seems OK to me:
1) Created a proxy class that in some way mimics the current behavior of the `previous_model` method.
```ruby
class PreviousModel
  attr_accessor :obj

  def initialize(obj)
    self.obj = obj
  end

  def changes_method_name
    ::CounterCulture::Counter::ACTIVE_RECORD_VERSION >= Gem::Version.new("5.1.0") ? :saved_changes : :changed_attributes
  end

  def changes_method
    @changes_method ||= obj.public_send(changes_method_name)
  end

  def method_missing(m, *args, &block)
    if m.in?(changes_method)
      value = changes_method[m]
      ::CounterCulture::Counter::ACTIVE_RECORD_VERSION >= Gem::Version.new("5.1.0") ? value.first : value
    else
      obj.send(m, *args, &block)
    end
  end
end
```

2) And replace the `previous_model` method with a simple instantiation of the previous class:
```ruby
def previous_model(obj)
  CounterCulture::PreviousModel.new(obj)
end
```

This solution avoids the `dup` & assign strategy that is currently in-place, keeps all the object methods accessible, and just the attributes (which are also methods) that were changed by the `changes_method` are intercepted and the previous value is returned.

Feedback appreciated :)

**Comment by magnusvk** [comment]
It's the first time I'm hearing of this issue, so it's probably not super common. That being said, I quite like your proxy class as an approach here. It strikes me as more elegant than having to `dup` the model. I'd be happy to look at a PR that introduces this.

**Comment by jfloff** [comment]
Sadly, this approach does not work 😅 It fails in scenarios where the counter cache has dynamic column names. The proxy class works for methods that the `counter_culture` lib calls directly on the `previous_model` object, but in the case of the dynamic column names we have calls from within the object itself. The `PolyImage` class [has a good example of that](https://github.com/magnusvk/counter_culture/blob/master/spec/models/poly_image.rb#L13): the `special?` method is intercepted, but anything within it is not, and due to that we are unable to get the previous result of that call.

I am exploring some alternatives, I will update you as I make progress :) If you have any ideas for how we could move away from the dup & assign you can throw them my way.

**Comment by jfloff** [comment]
Here is my first attempt. From my last comment, I think we need to keep the `dup` since that is the only way we could access the object and call the previous dynamic column parts. So my first thought was -- maybe I could just tweak the assign part.

After testing against the after_commit gem, this is the solution I came up with. 
Replace `prev[key] = old_value` with `prev.instance_variable_get(:@attributes).write_from_user(key, old_value)`.
In short, we are going as low as needed in the _write_ stack to bypass any gems that piggyback on ActiveRecord's `write_attribute` method, and change to the old values straight in the variable that that method manipulates.

**Comment by magnusvk** [comment]
Assuming that works across Rails versions, I'd be ok with that yeah. 😄
